goalie/mesh_seq.py

```diff
@@ -684,92 +468,134 @@ def solutions(self):
             self._create_solutions()
         return self._solutions
 
+    def _reinitialise_fields(self, initial_conditions):
+        """
+        Reinitialise fields and assign initial conditions on the given subinterval.
+
+        :arg initial_conditions: the initial conditions to assign
+        :type initial_conditions: :class:`dict` with :class:`str` keys and
+            :class:`firedrake.function.Function` values
+        """
+        for field in self.fields:
+            fs = initial_conditions[field].function_space()
+            self.fields[field] = (
+                firedrake.Function(fs, name=field),
+                firedrake.Function(fs, name=f"{field}_old"),
+            )
+            self.fields[field][1].assign(initial_conditions[field])
```

Comment:
This is really nice!

Formatting: you can use `for field, initial_condition in enumerate(initial_conditions):`.